packages/snaps-execution-environments/src/webworker/pool/WebWorkerPool.ts

```diff
@@ -0,0 +1,208 @@
+
+    this.#stream.on('data', this.#onData.bind(this));
+  }
+
+  /**
+   * Handle an incoming message from the `OffscreenExecutionService`. This
+   * assumes that the message contains a `jobId` property, and a JSON-RPC
+   * request in the `data` property.
+   *
+   * @param data - The message data.
+   * @param data.data - The JSON-RPC request.
+   * @param data.jobId - The job ID.
+   */
+  #onData(data: { data: JsonRpcRequest; jobId: string }) {
+    const { jobId, data: request } = data;
+
+    const job = this.jobs.get(jobId);
+    if (!job) {
+      // This ensures that a job is initialized before it is used. To avoid
+      // code duplication, we call the `#onData` method again, which will
+      // run the rest of the logic after initialization.
+      this.#initializeJob(jobId)
+        .then(() => {
+          this.#onData(data);
+        })
+        .catch((error) => {
+          logError('[Worker] Error initializing job:', error.toString());
+
+          this.#stream.write({
+            jobId,
+            data: {
+              name: 'command',
+              data: {
+                jsonrpc: '2.0',
+                id: request.id ?? null,
+                error: {
+                  code: -32000,
+                  message: 'Internal error',
+                },
+              },
+            },
+          });
+        });
+
+      return;
+    }
+
+    // This is a method specific to the `WebWorkerPool`, as the service itself
+    // does not have access to the iframes directly.
+    if (request.method === 'terminateJob') {
+      this.#terminateJob(jobId);
+      return;
+    }
+
+    job.stream.write(request);
+  }
+
+  /**
+   * Create a new iframe and set up a stream to communicate with it.
+   *
+   * @param jobId - The job ID.
+   * @returns The job.
+   */
+  async #initializeJob(jobId: string): Promise<ExecutorJob> {
+    const worker = await this.#getWorker();
+    const jobStream = new WebWorkerParentPostMessageStream({
+      worker,
+    });
+
+    // Write messages from the worker to the parent, wrapped with the job ID.
+    jobStream.on('data', (data) => {
+      this.#stream.write({ data, jobId });
+    });
+
+    const job = { id: jobId, worker, stream: jobStream };
+    this.jobs.set(jobId, job);
+    return job;
+  }
+
+  /**
+   * Terminate the job with the given ID. This will close the iframe and delete
+   * the job from the internal job map.
+   *
+   * @param jobId - The job ID.
+   */
+  #terminateJob(jobId: string) {
+    const job = this.jobs.get(jobId);
+    assert(job, `Job "${jobId}" not found.`);
+
+    job.stream.destroy();
+    job.worker.terminate();
+
+    this.jobs.delete(jobId);
+  }
+
+  /**
+   * Get a worker from the pool. A new worker will be created automatically.
+   *
+   * @returns The worker.
+   */
+  async #getWorker() {
+    // Lazily create the pool of workers.
+    if (this.pool.length === 0) {
+      await this.#updatePool();
+    }
+
+    const worker = this.pool.shift();
```

Comment:
This way our web workers are only ever used once. Do we want to experiment with re-using them in the pool?

Reply:
Potentially something we can do as a follow up to this PR? It's in a working state right now and I'd like to get this merged.

Reply:
Yeah, that's fine 👍